### PR TITLE
Fix SimpleFSDP metadata for uneven composed sharding

### PR DESCRIPTION
**Pull request by SongyuanZhao**
SimpleFSDP uses full global tensor metadata when applying DP sharding to a TP/EP-local tensor. It also relies on implicit shape inference when rewrapping DTensors, which cannot recover uneven shard sizes correctly. For example, a TP-local dimension of size 5 split across two DP ranks produces shards of size 3 and 2; inferring the global size from either shard gives 6 or 4 instead of 5.

  This change:
  - Uses the TP/EP-local tensor's shape and stride for the temporary DP redistribution specs.
  - Derives the DP-only logical shape and stride from the input DTensor metadata and passes them explicitly during rewrapping.
  - Restores the input's global shape and stride when wrapping the gathered tensor on the TP/EP mesh.
  - Preserves `requires_grad` when recreating parameters, so frozen parameters remain frozen.

  Adds 14 component regression cases to the existing `TestSimpleFSDP` class, covering uneven and empty shards, FSDP/HSDP with TP, EP+TP composition, transposed parameters, compilation, meta initialization, the `spmd_types` backend, and frozen parameters. The shared numerical checks compare outputs, losses, accumulated gradients, parameter updates, and AdamW state against an unsharded reference.

  Validation:
  - All 14 added cases passed using an 8-process CPU/Gloo launcher adapter running the original test methods.
  - File-scoped pre-commit checks passed, excluding Pyrefly due to existing repository-wide failures.

### CODE REVIEW

**Comment by tianyu-l** [comment]
@claude please verify if this is a legit problem. If so, does it happen with `spmd_backend` being "partial_dtensor" or "spmd_types" (introduced in https://github.com/pytorch/torchtitan/pull/4325) or both?

**Comment by tianyu-l** [comment]
@claude partial_dtensor backend is being deprecated so I don't care about that at all. For spmd_types, provide a simple example to showcase where the bug happens, and how this PR would fix it, to help me
1. understand the nature of the bug
2. evaluate if this is the best fix possible

File: torchtitan/experiments/graph_trainer/simple_fsdp.py [hunk 1]
```diff
@@ -346,7 +365,11 @@ def replicate_compute(self, x: DTensor) -> torch.Tensor:
                         )
             else:
                 output = DTensor.from_local(
```
Comment (tianyu-l) [inline torchtitan/experiments/graph_trainer/simple_fsdp.py]: this branch is removed in https://github.com/pytorch/torchtitan/pull/4419

File: torchtitan/experiments/graph_trainer/simple_fsdp.py [hunk 2]
```diff
@@ -303,10 +311,27 @@ def replicate_compute(self, x: DTensor) -> torch.Tensor:
...
+            # shards. DP-local sizes alone cannot recover uneven shard lengths.
+            local_shape, _ = compute_local_shape_and_global_offset(
+                x.shape, non_dp_mesh, non_dp_placements, skip_offset=True
+            )
             # re-wrap 2D DTensor to 1D DTensor on dp_mesh for efficient FSDP all-gather
             sharded_local_tensor = x.to_local()
+            # Preserve the input's logical dimension order when projecting its
+            # layout onto the TP/EP-local shape.
+            local_stride = compute_local_stride(x.stride(), local_shape)
             sharded_dtensor = DTensor.from_local(
```
Comment (tianyu-l) [inline torchtitan/experiments/graph_trainer/simple_fsdp.py:320]: @pianpwk this seems a legit bug for DP-uneven sharding, assuming TP / EP-even sharding.

File: torchtitan/experiments/graph_trainer/simple_fsdp.py
Comment (tianyu-l) [inline torchtitan/experiments/graph_trainer/simple_fsdp.py:113]: what about TP / EP uneven sharding?

Comment (SongyuanZhao) [reply torchtitan/experiments/graph_trainer/simple_fsdp.py:113]: The current spmd_types path still has a gap for uneven TP/EP sharding. The input here is already a local tensor, so the original global shape is unavailable. Supporting this would require broader changes to preserve and pass that metadata.

I also found that Module.parallelize() already rejects uneven TP/EP parameter sharding.

Would you prefer to address this separately, or extend this PR beyond the uneven-DP fix?

File: torchtitan/experiments/graph_trainer/simple_fsdp.py (hunk 1 above)
Comment (SongyuanZhao) [reply torchtitan/experiments/graph_trainer/simple_fsdp.py]: Would you prefer that I drop the changes to this branch from this PR?

Comment (SongyuanZhao) [reply torchtitan/experiments/graph_trainer/simple_fsdp.py]: Rebased onto the latest main and dropped the changes to the branch removed by #4419.

File: torchtitan/experiments/graph_trainer/simple_fsdp.py
Comment (tianyu-l) [reply torchtitan/experiments/graph_trainer/simple_fsdp.py:113]: > Module.parallelize() already rejects uneven TP/EP parameter sharding

sounds good -- as long as it is rejected, it's good enough

**Review by pianpwk — APPROVED** [review]
looks good, thank you for the fix

**Review by tianyu-l — APPROVED** [review]
